### PR TITLE
KEYCLOAK-18051 Add regex redirect-uri client policy

### PR DESCRIPTION
**Pull request by patrick-primesign**
Documentation-PR: https://github.com/keycloak/keycloak-documentation/pull/1223

### CODE REVIEW

**Comment by stianst** [comment]
Sorry for the late reply to this one. Wondering if there's still interest in this PR if so perhaps we can discuss how to proceed.

If so could you provide a bit more context behind the approach here? It seems like the idea is that clients can register '*' as redirect-uris, then the client-policy is used to configure a regex-pattern that can check the redirect-uri used by the authorization request. That seems a bit the wrong way around to me.

I would have rather expected to have a client-policy that is used to decide what redirect-uris a client can register, so would check on client create/update rather than on authz requests.

Wondering if it also might be nice to convert this into a redirect-uri policy and have some more options than just regex to incorporate some of the stuff we've discussed in https://github.com/keycloak/keycloak/discussions/9278.

**Comment by stianst** [comment]
One more thing is that it would be good to separate the client-ids condition into a separate PR.

**Comment by stianst** [comment]
Added some more thoughts to https://github.com/keycloak/keycloak/discussions/9278#discussioncomment-3606032

**Comment by CorneliaLahnsteiner** [comment]
Hi @stianst. There is still interest in bringing this pull request upstream. We are already using this change in production (since a year) and it works pretty well. 

The idea behind this PR is to allow (next to wildcards) RegEx for the allowed redirect-uri. However, the specs very clearly state that redirect-uris should be a strict string comparison. Therefore, we decided that an realm admin has to explicitly configure the RegEx by a client policy. 

Configuration steps: 
1. Register 'https://*' as valid redirect-uris at client level
2. Register in realm settings under client policies a new profile and add the executor "regex-redirect-uri" where you configure the valid redirect-uris in RegEx pattern
3. Register a new client policy and add the condition "client-ids" where you configure the clients who should be affected by the RegEx validation
4. Finally add the client profile to the client policy to enable the RegEx validation for the specific clients

The RegEx pattern is used by the authorization request to validate the redirect-uri.

To understand your suggestion: Your approach would be to register a client policy that decides what kind of valid-redirect uris (wildcard, RegEx) a client may have. Furthermore, adapt the client settings to also allow RegEx (if a client-policy is set for the specific client). And change this method to also support RegEx validation: https://github.com/keycloak/keycloak/blob/ae7c01b7195574cb7a1d59bf76cfbe95f6d1bda0/services/src/main/java/org/keycloak/protocol/oidc/utils/RedirectUtils.java#L178
Your approach also seems justified to me, but it is a deeper change in Keycloak than the current change. A question to your approach: At the client settings, can we query if this client has set the regex client policy? What should happen, if we remove this client at the client-policies? 

Thanks,
Cornelia

**Comment by CorneliaLahnsteiner** [comment]
> One more thing is that it would be good to separate the client-ids condition into a separate PR.

It is no proplem to seperate the condition into a sperate PR.

**Comment by stianst** [comment]
There's a more extensive executor in https://github.com/keycloak/keycloak/pull/14659. Perhaps we would want to keep that one, and in this one only keep the client-id condition?

File: themes/src/main/resources/theme/base/admin/messages/admin-messages_en.properties [hunk 1]
```diff
@@ -904,6 +904,8 @@ client-updater-trusted-hosts.label=Trusted hosts
 client-updater-trusted-hosts.tooltip=List of Hosts, which are trusted. In case that client registration/update request comes from the host/domain specified in this configuration, condition evaluates to true. You can use hostnames or IP addresses. If you use star at the beginning (for example '*.example.com' ) then whole domain example.com will be trusted.
 client-updater-source-roles.label=Updating entity role
 client-updater-source-roles.tooltip=The condition is checked during client registration/update requests and it evaluates to true if the entity (usually user), who is creating/updating client is member of the specified role. For reference the realm role, you can use the realm role name like 'my_realm_role' . For reference client role, you can use the client_id.role_name for example 'my_client.my_client_role' will refer to client role 'my_client_role' of client 'my_client'.
+client-ids.label=Client-IDs
```
Comment (jonkoops) [inline themes/src/main/resources/theme/base/admin/messages/admin-messages_en.properties:907]: Are these new labels consumed anywhere?

Comment (1-alex98) [reply themes/src/main/resources/theme/base/admin/messages/admin-messages_en.properties:907]: yes
ClientIdsConditionFactory Line 20

**Comment by mposolda** [comment]
Few points to this:
- There was executor added in the PR for https://github.com/keycloak/keycloak/issues/25637 and should be available in Keycloak 24. Hence the "executor" part of this PR is out-of-date now and can be likely removed

- So effectively the only "new" part to this PR might be condition for client-ids. But I am wondering if this condition is really needed? As for instance you can use existing condition `ClientRolesCondition`, which applies when client has client role with the particular name. This allows that you can add client role to all requested clients (this role can be used purely as "marker role" to specify that particular client policy should be applied. No users need to be added to this role at all). It seems the approach based on "client role" has more flexibility and it can scale much better as in case when realm has thousands of clients and you want policy to be executed for half of them for instance, it won't scale well to have client-ids condition with thousand of client-ids explicitly named. It would be bad for both performance (iterating over 1000 client-ids in every request) as well as manageability.

**Comment by CorneliaLahnsteiner** [comment]
> Few points to this:
> 
> * There was executor added in the PR for [Client policies: executor for validate and match a redirect URI #25637](https://github.com/keycloak/keycloak/issues/25637) and should be available in Keycloak 24. Hence the "executor" part of this PR is out-of-date now and can be likely removed
> * So effectively the only "new" part to this PR might be condition for client-ids. But I am wondering if this condition is really needed? As for instance you can use existing condition `ClientRolesCondition`, which applies when client has client role with the particular name. This allows that you can add client role to all requested clients (this role can be used purely as "marker role" to specify that particular client policy should be applied. No users need to be added to this role at all). It seems the approach based on "client role" has more flexibility and it can scale much better as in case when realm has thousands of clients and you want policy to be executed for half of them for instance, it won't scale well to have client-ids condition with thousand of client-ids explicitly named. It would be bad for both performance (iterating over 1000 client-ids in every request) as well as manageability.

@mposolda great, thanks for the information! I will have a look at the already merged solution (https://github.com/keycloak/keycloak/issues/25637) in the next few days. If our solution is outdated by the solution already merged, then this pull request offers no added value and can be closed. 
And I agree with your second comment. It is more performant to use the already existing client roles instead of our client-ids.

**Comment by mposolda** [comment]
@CorneliaLahnsteiner Cool, Thanks! Please comment here or close the PR if you are already comfortable with existing approach. Thanks for your contribution regardless!

**Comment by mposolda** [comment]
@CorneliaLahnsteiner @patrick-primesign No further feedback, so I am closing this for now. Just to summary:
- Regex redirect-uri enforcer executor is already available in Keycloak 24
- Condition for client-ids might not be needed as condition for `client-roles` can be used

But feel free to re-open if you think that some parts of this PR still should be added (or maybe rather create new GH issue and/or PR if something doesn't work as expected in Keycloak 24, which is going to be released soon).

**Pull request closed without merging**